galite-core/src/main/kotlin/org/kopi/galite/form/VBlock.kt

```diff
@@ -1420,17 +1421,141 @@ abstract class VBlock(var form: VForm) : VConstants, DBContextHandler, ActionHan
+      if (fields[i].getColumnCount() > 0) {
+        idqry += 1
+      }
+    }
+
+    // open database query, fetch tuples
+    val query = table.slice(columns!!).select(condition).orderBy(*orderBy.toTypedArray())
+
+    fetchCount = 0
+    for (result in query) {
+      if (fetchCount < fetchSize) {
+        if (result[columns[idqry]] == 0) {
+          continue
+        }
+        fetchBuffer[fetchCount] = result[columns[idqry]] as Int
+        if (fetchCount >= bufferSize) {
+          fetchCount += 1
+        } else {
+          var j = 0
+
+          fields.forEach { field ->
+            if (field.getColumnCount() > 0) {
+              field.setQuery_(fetchCount, result, columns[j])
+              j += 1
+            }
+          }
+          setRecordFetched(fetchCount, true)
+          setRecordChanged(fetchCount, false)
+          setRecordDeleted(fetchCount, false)
+          try {
+            if (isMulti()) {
+              activeRecord = fetchCount
+            }
+            callProtectedTrigger(VConstants.TRG_POSTQRY)
+            if (isMulti()) {
+              activeRecord = -1
+            }
+            fetchCount += 1
+          } catch (e: VException) {
+            if (isMulti()) {
+              activeRecord = -1
+            }
+            if (e is VSkipRecordException) {
+              clearRecordImpl(fetchCount)
+            } else {
+              clear()
+              throw e
+            }
+          } catch (t: Throwable) {
+            t.printStackTrace()
+          }
+        }
+      }
+    }
+
+    fetchPosition = 0
+    // !!! REMOVE setActiveRecord(0);
+    if (!isMulti() && fetchCount == 0) {
+      throw VQueryNoRowException(MessageCode.getMessage("VIS-00022"))
+    } else if (!isMulti()) {
+      setMode(VConstants.MOD_UPDATE)
+    }
+    fireBlockChanged()
   }
 
   /**
    * Fetches record with given ID from database.
    * @exception VException      an exception may be raised by triggers
    */
-  fun fetchRecord(id: Int) {
-    TODO()
+  fun fetchRecord(id: Int)  {
+    val columns = getSearchColumns()
+    val table = getSearchTables_()
+    val condition = mutableListOf<Op<Boolean>>()
+
+    condition.add(Op.build { idColumn eq id  })
+    if (VBlockDefaultOuterJoin.getFetchRecordCondition(fields) != null) {
+      condition.add(VBlockDefaultOuterJoin.getFetchRecordCondition(fields)!!)
+    }
+
+    val query = table.slice(columns!!).select(condition.compoundAnd())
+
+    transaction {
+      if (!query.execute(this)!!.next()) {
+        /* Record does not exist anymore: it was deleted by another user */
+        throw VSkipRecordException()
+      } else {
+        /* set values */
+        var j = 0
+
+        for (field in fields) {
+          if (field.getColumnCount() > 0) {
+            for (result in query) {
+              field.setQuery_(result, columns[j])
+            }
+            j += 1
+          }
+        }
+        assert(!query.execute(this)!!.next()) { "too many rows" }
+      }
```

Comment:
use single() method, and catch `NoSuchElementException` and `IllegalArgumentException` seperately with corresponding error message